### PR TITLE
Implementation of IdentityServer4

### PR DESCRIPTION
**Pull request by jpw547-zz**
The basis for this implementation comes from following [this blog post](https://fullstackmark.com/post/21/user-authentication-and-identity-with-angular-aspnet-core-and-identityserver). Nearly 100% of the code written here is identical to the [source code](https://github.com/mmacneil/AngularASPNETCoreOAuth/tree/master/src/AuthServer) for that post. Minor changes were made, such as the cookie policy in Startup.cs on line 192, which was actually necessary for the redirection to work with our Angular application.

### CODE REVIEW

**Comment by jpw547-zz** [comment]
Pipeline stuff is in the kubernetes branch and currently being worked on by @ibnetahir1

File: src/AuthServer/AuthServer.Infrastructure/AuthServer.Infrastructure.csproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="IdentityServer4.EntityFramework" Version="4.1.2" />
+    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="5.0.5" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.5" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="5.0.5">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.5" />
```
Comment (escalonn) [inline src/AuthServer/AuthServer.Infrastructure/AuthServer.Infrastructure.csproj:15]: sqlserver consumes more resources for local testing, but if you have a reason to prefer it to pgsql... then ok.

File: src/AuthServer/AuthServer.Infrastructure/Data/DesignTimeDbContextFactoryBase.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.IO;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Design;
+using Microsoft.Extensions.Configuration;
+
+namespace AuthServer.Infrastructure.Data
+{
+    public abstract class DesignTimeDbContextFactoryBase<TContext> : IDesignTimeDbContextFactory<TContext> where TContext : DbContext
```
Comment (escalonn) [inline src/AuthServer/AuthServer.Infrastructure/Data/DesignTimeDbContextFactoryBase.cs:9]: why are we doing this? doesn't asp.net core have this built-in via the CreateDefaultBuilder call? isn't that just as accessible to dotnet ef as in our other services?

File: src/AuthServer/AuthServer.Infrastructure/Data/Identity/PersistedGrantDbContextFactory.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+using Microsoft.EntityFrameworkCore.Design;
+
+namespace AuthServer.Infrastructure.Data.Identity
+{
+    public class PersistedGrantDbContextFactory : IDesignTimeDbContextFactory<PersistedGrantDbContext>
+    {
+        public PersistedGrantDbContext CreateDbContext(string[] args)
+        {
+            var optionsBuilder = new DbContextOptionsBuilder<PersistedGrantDbContext>();
+            optionsBuilder.UseSqlServer("Server=(localdb)\\mssqllocaldb;Database=AuthServer;Trusted_Connection=True;MultipleActiveResultSets=true",
```
Comment (escalonn) [inline src/AuthServer/AuthServer.Infrastructure/Data/Identity/PersistedGrantDbContextFactory.cs:14]: localdb? that will definitely not work in production

**Review by escalonn — CHANGES_REQUESTED** [review]
this PR is too large. you should have seen this coming and prepared it better. what should have happened:
1. PR with bare minimum template .net projects, with pipeline, docker, k8s, etc.
2. PRs with any additional tool-generated code.
3. PRs with your own modifications.

i guess it's too late now for that, but as it is i don't know what code here came from templates and what code you wrote. write a PR description giving as much context to the code review as possible, including what references you followed online. you never want to leave that kind of thing out of the comments/PR documentation, it's irreplaceable when you or someone else comes back to the code months later and can't remember what's going on. the PR description should identify which folders and files are handwritten or modified, as well as what steps you followed to generate the rest of the code.

File: src/AuthServer/AuthServer.Infrastructure/AuthServer.Infrastructure.csproj (hunk 1 above)
Comment (jpw547-zz) [reply src/AuthServer/AuthServer.Infrastructure/AuthServer.Infrastructure.csproj:15]: No real reasoning for using SqlServer other than that's how to the blog post we were following did it. We can change it to pgsql

File: src/AuthServer/AuthServer.Infrastructure/Data/DesignTimeDbContextFactoryBase.cs (hunk 2 above)
Comment (jpw547-zz) [reply src/AuthServer/AuthServer.Infrastructure/Data/DesignTimeDbContextFactoryBase.cs:9]: This file was part of the source code that we got from the blog post implementation, so we didn't actually write this.

**Pull request merged**